constantine/signatures/ecdsa.nim

```diff
@@ -0,0 +1,311 @@
+    # `r = k·G (mod n)`
+    R.scalarMul(k, G)
+    # get x coordinate of the point `r` *in affine coordinates*
+    let rx = R.getAffine().x
+    let r = Fr[Name].fromBig(rx.toBig()) # convert to `Fr`
+
+    if bool(r.isZero()):
+      continue # try again
+
+    # Calculate s
+    # `s = (k⁻¹ · (h + r · p)) (mod n)`
+    # with `h`: message hash as `Fr[C]` (if we didn't use SHA256 w/ 32 byte output
+    # we'd need to truncate to N bits for N being bits in modulo `n`)
+    var s {.noinit.}: Fr[Name]
+    s.prod(r, secretKey) # `r * secretKey`
+    s += message_hash     # `message_hash + r * secretKey`
+    k.inv()               # `k := k⁻¹`
+    s *= k                # `k⁻¹ * (message_hash + r * secretKey)`
+    # get inversion of `s` for 'lower-s normalization'
+    var sneg = s # inversion of `s`
+    sneg.neg()   # q - s
+    # conditionally assign result based on BigInt comparison
+    let mask = s.toBig() > sneg.toBig() # if true, `s` is in upper half, need `sneg`
+    ccopy(s, sneg, mask)
+
+    if bool(s.isZero()):
+      continue # try again
+
+    # Set output and return
+    sig.r = r
+    sig.s = s
+    return
+
+proc coreSign*[Sig, SecKey](
+    signature: var Sig,
+    secretKey: SecKey,
+    message: openArray[byte],
+    H: type CryptoHash,
+    nonceSampler: NonceSampler = nsRandom) {.genCharAPI.} =
+  ## Computes a signature for the message from the specified secret key.
+  ##
+  ## Output:
+  ## - `signature` is overwritten with `message` signed with `secretKey`
+  ##
+  ## Inputs:
+  ## - `Hash` a cryptographic hash function.
+  ##   - `Hash` MAY be `sha256`
+  ##   - `Hash` MAY be `keccak`
+  ##   - Otherwise, H MUST be a hash function that has been proved
+  ##    indifferentiable from a random oracle [MRH04] under a reasonable
+  ##    cryptographic assumption.
+  ## - `message` is the message to hash
+  signature.signImpl(secretKey, message, H, nonceSampler)
+
+proc verifyImpl[Name: static Algebra; Sig](
+    publicKey: EC_ShortW_Aff[Fp[Name], G1],
+    signature: Sig, # tuple[r, s: Fr[Name]],
+    message: openArray[byte],
+    H: type CryptoHash,
+): bool =
+  ## Verify a given `signature` for a `message` using the given `publicKey`.
+  # 1. Hash the message (same as in signing)
+  var dgst {.noinit.}: array[H.digestSize, byte]
+  H.hash(dgst, message)
+  var e {.noinit.}: Fr[Name]
+  e.fromDigest(dgst, truncateInput = true)
+
+  # 2. Compute w = s⁻¹
+  var w = signature.s
+  w.inv() # w = s⁻¹
+
+  # 3. Compute u₁ = ew and u₂ = rw
+  var
+    u1 {.noinit.}: Fr[Name]
+    u2 {.noinit.}: Fr[Name]
+  u1.prod(e, w)
+  u2.prod(signature.r, w)
+
+  # 4. Compute u₁G + u₂Q
+  var
+    point1 {.noinit.}: EC_ShortW_Jac[Fp[Name], G1]
+    point2 {.noinit.}: EC_ShortW_Jac[Fp[Name], G1]
+  # Generator of the curve
+  const G = publicKey.F.Name.getGenerator($publicKey.G)
+  point1.scalarMul(u1, G)
+  point2.scalarMul(u2, publicKey)
+  var R {.noinit.}: EC_ShortW_Jac[Fp[Name], G1]
+  R.sum(point1, point2)
```

Comment:
As mentioned in https://github.com/mratsim/constantine/pull/504#discussion_r1900762614 this should be a dedicated proc, which would be subject to optimization in a later PR as it's a bottleneck for blockchain (signatures are verified all the time) and TLS.

Reply:
Yup. I'll open an issue about these two cases (here and the other PR).